# EDA 

In [2]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
import sqlite3 as sql

## 1. Cargando el dataset

In [3]:
ruta_archivo = "../data/raw/adidas_us_sales_dataset.xlsx"

df = pd.read_excel(ruta_archivo, index_col=None)

# Eliminar columna de índice anterior si existe
if 'Unnamed: 0' in df.columns:
    df.drop(columns=['Unnamed: 0'], inplace=True)

# Guardar en SQLite
conn = sql.connect("../data/raw/adidas_sales.db")
df.to_sql("ventas_adidas", conn, if_exists="replace", index=False)

print("Base de datos creada con éxito")

cursor = conn.cursor()


Base de datos creada con éxito


## 2. Conociendo la estructura del dataset

In [4]:
# Número de filas y columnas:
print(f"Dimensiones del dataset: {df.shape}")

# Información sobre los tipos de datos
print("\nInformación del dataset:")
df.info()

Dimensiones del dataset: (9648, 13)

Información del dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9648 entries, 0 to 9647
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Retailer          9648 non-null   object        
 1   Retailer ID       9648 non-null   int64         
 2   Invoice Date      9648 non-null   datetime64[ns]
 3   Region            9648 non-null   object        
 4   State             9648 non-null   object        
 5   City              9648 non-null   object        
 6   Product           9648 non-null   object        
 7   Price per Unit    9648 non-null   float64       
 8   Units Sold        9648 non-null   int64         
 9   Total Sales       9648 non-null   float64       
 10  Operating Profit  9648 non-null   float64       
 11  Operating Margin  9648 non-null   float64       
 12  Sales Method      9648 non-null   object        
dtypes: datetime64[ns

### 🔍 Análisis:

- Tenemos 9648 filas y 13 columnas.
- Hay variables numéricas (float64, int64) y categóricas(object).
- No hay valores nulos en el dataset.

## 3. Revisando los nombres de las columnas

In [6]:
# Nombres de las columnas
print(f"Nombres de las columnas:")
df.columns.tolist()

Nombres de las columnas:


['Retailer',
 'Retailer ID',
 'Invoice Date',
 'Region',
 'State',
 'City',
 'Product',
 'Price per Unit',
 'Units Sold',
 'Total Sales',
 'Operating Profit',
 'Operating Margin',
 'Sales Method']

## 4. Tipos de variables detectadas

In [7]:
print("Tipos de variables en el dataset:")
print(df.dtypes)

Tipos de variables en el dataset:
Retailer                    object
Retailer ID                  int64
Invoice Date        datetime64[ns]
Region                      object
State                       object
City                        object
Product                     object
Price per Unit             float64
Units Sold                   int64
Total Sales                float64
Operating Profit           float64
Operating Margin           float64
Sales Method                object
dtype: object


### 🔍 Interpretación:

- Numéricas: `Retailer ID`, `Price per Unit`, `Units Sold`, `Total Sales`, `Operating Profit`, `Operating Margin`.
- Categóricas: `Retailer`, `Region`, `State`, `City`, `Product`, `Sales Method`.
- Fechas: `Invoice Date`.

## 5. Análisis de la Distribución de Datos

- Análisis descriptivo rápido para ver la distribución de los valores:

In [12]:
# Para variables numéricas
df.describe()

Retailer ID  Price per Unit   Units Sold    Total Sales  \
count  9.648000e+03     9648.000000  9648.000000    9648.000000   
mean   1.173850e+06       45.216625   256.930037   93273.437500   
std    2.636038e+04       14.705397   214.252030  141916.016727   
min    1.128299e+06        7.000000     0.000000       0.000000   
25%    1.185732e+06       35.000000   106.000000    4254.500000   
50%    1.185732e+06       45.000000   176.000000    9576.000000   
75%    1.185732e+06       55.000000   350.000000  150000.000000   
max    1.197831e+06      110.000000  1275.000000  825000.000000   

       Operating Profit  Operating Margin  
count       9648.000000       9648.000000  
mean       34425.244761          0.422991  
std        54193.113713          0.097197  
min            0.000000          0.100000  
25%         1921.752500          0.350000  
50%         4371.420000          0.410000  
75%        52062.500000          0.490000  
max       390000.000000          0.800000

In [13]:
# Para variables categóricas
df.describe(include=['object', 'category', 'bool'])

Retailer Region       State      City                Product  \
count          9648   9648        9648      9648                   9648   
unique            6      5          50        52                      6   
top     Foot Locker   West  California  Portland  Men's Street Footwear   
freq           2637   2448         432       360                   1610   

       Sales Method  
count          9648  
unique            3  
top          Online  
freq           4889

### 🔍 Análisis :

- Para variables numéricas: Nos interesa la media (`mean`), mediana (`50%`), valores extremos (`min` y `max`) para cada variable o columna.

- Para variables categóricas: Vemos cuántas categorías existen (`unique`) y cuál es el más frecuente (`top`).


#### Valores distintos `unique`:

- Retailer: 6 valores (categorías) únicos(as).
- Region: 5 valores (categorías) únicos(as).
- State: 50 valores (categorías) únicos(as).
- City: 52 valores (categorías) únicos(as).
- Product: 6 valores (categorías) únicos(as).
- Sales Method: 3 valores (categorías) únicos(as).

#### Valores más frecuentes (moda de las variables categóricas)

- Retailer: 
    - count: Hay 9648 datos.
    - unique: Hay 6 valores diferentes o únicos.
    - top: El dato más común es `Foot Locker`.
    - freq: El dato más común aparece 2637 veces.

- Region:
    - count: Hay 9648 datos.
    - unique: Hay 5 valores diferentes o únicos.
    - top: El dato más común es `West`.
    - freq: El dato más común aparece 2448 veces.

- State:
    - count: Hay 9648 datos.
    - unique: Hay 50 valores diferentes o únicos.
    - top: El dato más común es `California`.
    - freq: El dato más común aparece 432 veces.

- City:
    - count: Hay 9648 datos.
    - unique: Hay 52 valores diferentes o únicos.
    - top: El dato más común es `Portland`.
    - freq: El dato más común aparece 360 veces.

- Product:
    - count: Hay 9648 datos.
    - unique: Hay 6 valores diferentes o únicos.
    - top: El dato más común es `Men's Street Footwear`.
    - freq: El dato más común aparece 1610 veces.

- Sales Method:
    - count: Hay 9648 datos.
    - unique: Hay 3 valores diferentes o únicos.
    - top: El dato más común es `Online`.
    - freq: El dato más común aparece 4889 veces.

## 6. Exploración de calidad de datos

### ✅ Datos Faltantes o Nulos

In [20]:
print("Valores nulos por columna:")
df.isnull().sum()

Valores nulos por columna:


Retailer            0
Retailer ID         0
Invoice Date        0
Region              0
State               0
City                0
Product             0
Price per Unit      0
Units Sold          0
Total Sales         0
Operating Profit    0
Operating Margin    0
Sales Method        0
dtype: int64

In [21]:
# Usando SQLite
valores_nulos = """
                SELECT
                    SUM(CASE WHEN Retailer IS NULL THEN 1 ELSE 0 END) AS retailer_nulos,
                    SUM(CASE WHEN `Retailer ID` IS NULL THEN 1 ELSE 0 END) AS retailerid_nulos,
                    SUM(CASE WHEN `Invoice Date` IS NULL THEN 1 ELSE 0 END) AS invoice_date_nulos,
                    SUM(CASE WHEN Region IS NULL THEN 1 ELSE 0 END) AS region_nulos,
                    SUM(CASE WHEN State IS NULL THEN 1 ELSE 0 END) AS state_nulos,
                    SUM(CASE WHEN City IS NULL THEN 1 ELSE 0 END) AS city_nulos,
                    SUM(CASE WHEN Product IS NULL THEN 1 ELSE 0 END) AS products_nulos,
                    SUM(CASE WHEN `Price per Unit` IS NULL THEN 1 ELSE 0 END) AS price_per_unit_nulos,
                    SUM(CASE WHEN `Units Sold` IS NULL THEN 1 ELSE 0 END) AS units_sold_nulos,
                    SUM(CASE WHEN `Total Sales` IS NULL THEN 1 ELSE 0 END) AS total_sales_nulos,
                    SUM(CASE WHEN `Operating Profit` IS NULL THEN 1 ELSE 0 END) AS operating_profit_nulos,
                    SUM(CASE WHEN `Operating Margin` IS NULL THEN 1 ELSE 0 END) AS operating_margin_nulos,
                    SUM(CASE WHEN `Sales Method` IS NULL THEN 1 ELSE 0 END) AS sales_methond_nulos
                FROM ventas_adidas;
                """
valores_nulos = pd.read_sql_query(valores_nulos, conn)
valores_nulos

retailer_nulos  retailerid_nulos  invoice_date_nulos  region_nulos  \
0               0                 0                   0             0   

   state_nulos  city_nulos  products_nulos  price_per_unit_nulos  \
0            0           0               0                     0   

   units_sold_nulos  total_sales_nulos  operating_profit_nulos  \
0                 0                  0                       0   

   operating_margin_nulos  sales_methond_nulos  
0                       0                    0

### ✅ Datos Duplicados

In [22]:
duplicados = df.duplicated().sum()
print(f"Número de filas duplicadas: {duplicados}")

Número de filas duplicadas: 0


In [23]:
consulta = """
            WITH duplicados AS (
                SELECT *,
                    ROW_NUMBER() OVER (
                        PARTITION BY `Retailer ID`, `Invoice Date`, Product, `Sales Method`, City, State, `Price per Unit`
                        ORDER BY `Units Sold` DESC
                    ) AS num_fila
                FROM ventas_adidas
            )
            SELECT *
            FROM duplicados
            WHERE num_fila = 2;
            """
duplicados = pd.read_sql_query(consulta, conn)
duplicados

Retailer  Retailer ID                Invoice Date     Region  \
0       West Gear      1185732  1969-12-31 19:00:43.945000  Southeast   
1       West Gear      1185732  1969-12-31 19:00:43.945000  Southeast   
2     Foot Locker      1185732  1969-12-31 19:00:44.245000      South   
3   Sports Direct      1185732  1969-12-31 19:00:44.263000    Midwest   
4          Amazon      1185732  1969-12-31 19:00:44.264000    Midwest   
5   Sports Direct      1185732  1969-12-31 19:00:44.289000  Southeast   
6          Kohl's      1197831  1969-12-31 19:00:44.241000    Midwest   
7          Kohl's      1197831  1969-12-31 19:00:44.267000    Midwest   
8          Kohl's      1197831  1969-12-31 19:00:44.267000    Midwest   
9         Walmart      1197831  1969-12-31 19:00:44.274000      South   
10      West Gear      1197831  1969-12-31 19:00:44.274000      South   
11        Walmart      1197831  1969-12-31 19:00:44.274000      South   
12         Kohl's      1197831  1969-12-31 19:00:44.299000    Midwest   
13        Walmart      1197831  1969-12-31 19:00:44.306000      South   

             State           City                    Product  Price per Unit  \
0          Florida          Miami    Men's Athletic Footwear            55.0   
1          Florida          Miami    Men's Athletic Footwear            54.0   
2      Mississippi        Jackson  Women's Athletic Footwear            25.0   
3     North Dakota          Fargo  Women's Athletic Footwear            25.0   
4             Ohio       Columbus    Women's Street Footwear            20.0   
5   North Carolina      Charlotte    Women's Street Footwear            20.0   
6           Kansas        Wichita  Women's Athletic Footwear            20.0   
7           Kansas        Wichita      Men's Street Footwear            20.0   
8           Kansas        Wichita  Women's Athletic Footwear            20.0   
9         Arkansas    Little Rock    Men's Athletic Footwear            20.0   
10        Oklahoma  Oklahoma City  Women's Athletic Footwear            20.0   
11        Arkansas    Little Rock    Women's Street Footwear            10.0   
12          Kansas        Wichita    Men's Athletic Footwear            25.0   
13        Arkansas    Little Rock  Women's Athletic Footwear            20.0   

    Units Sold  Total Sales  Operating Profit  Operating Margin Sales Method  \
0          775     426250.0          149187.5              0.35     In-store   
1          194      10476.0            5238.0              0.50       Online   
2           91       2275.0            1137.5              0.50       Online   
3           22        550.0             231.0              0.42       Online   
4          124       2480.0            1314.4              0.53       Online   
5          144       2880.0            1584.0              0.55       Online   
6           42        840.0             378.0              0.45       Online   
7          153       3060.0            1499.4              0.49       Online   
8           26        520.0             249.6              0.48       Online   
9           98       1960.0             882.0              0.45       Online   
10          34        680.0             340.0              0.50       Online   
11          88        880.0             475.2              0.54       Online   
12          58       1450.0             696.0              0.48       Online   
13          63       1260.0             667.8              0.53       Online   

    num_fila  
0          2  
1          2  
2          2  
3          2  
4          2  
5          2  
6          2  
7          2  
8          2  
9          2  
10         2  
11         2  
12         2  
13         2

In [ ]:
# Query para ver si realmente hay registros duplicados
query = """
WITH duplicados AS (
    SELECT *,
        ROW_NUMBER() OVER (
            PARTITION BY `Retailer ID`, `Invoice Date`, Product, `Sales Method`, City, State, `Price per Unit`
            ORDER BY `Units Sold` DESC
        ) AS num_fila
    FROM ventas_adidas
),
seleccionados AS (
    SELECT *
    FROM duplicados
    WHERE num_fila IN (1, 2)
)
SELECT *
FROM (
    SELECT *
    FROM seleccionados
    WHERE num_fila = 1
    LIMIT 1
)
UNION ALL
SELECT *
FROM (
    SELECT *
    FROM seleccionados
    WHERE num_fila = 2
    LIMIT 1
);

"""
consulta = pd.read_sql_query(query, conn)
consulta


Retailer  Retailer ID                Invoice Date     Region  \
0  West Gear      1128299  1969-12-31 19:00:43.832000  Northeast   
1  West Gear      1185732  1969-12-31 19:00:43.945000  Southeast   

          State          City                  Product  Price per Unit  \
0  Pennsylvania  Philadelphia    Men's Street Footwear            32.0   
1       Florida         Miami  Men's Athletic Footwear            55.0   

   Units Sold  Total Sales  Operating Profit  Operating Margin Sales Method  \
0         122       3904.0           1991.04              0.51       Online   
1         775     426250.0         149187.50              0.35     In-store   

   num_fila  
0         1  
1         2

### 🔍 Interpretación:

- No hay valores duplicados, aunque la query de sql nos diga que hay 13 filas duplicadas, pero esto se puede ver que es cierto pero viendolo desde un punto de vista diferente

- Al detectar los registros duplicados con Python, vemos que hay 0 duplicados.

### ✅ Errores en Variables Categóricas

En este caso:
- Categóricas: `Retailer`, `Region`, `State`, `City`, `Product`, `Sales Method`.

In [32]:
df['Retailer'].unique()

array(['Foot Locker', 'Walmart', 'Sports Direct', 'West Gear', "Kohl's",
       'Amazon'], dtype=object)

In [34]:
df['Region'].unique()

array(['Northeast', 'South', 'West', 'Midwest', 'Southeast'], dtype=object)

In [35]:
df['State'].unique()

array(['New York', 'Texas', 'California', 'Illinois', 'Pennsylvania',
       'Nevada', 'Colorado', 'Washington', 'Florida', 'Minnesota',
       'Montana', 'Tennessee', 'Nebraska', 'Alabama', 'Maine', 'Alaska',
       'Hawaii', 'Wyoming', 'Virginia', 'Michigan', 'Missouri', 'Utah',
       'Oregon', 'Louisiana', 'Idaho', 'Arizona', 'New Mexico', 'Georgia',
       'South Carolina', 'North Carolina', 'Ohio', 'Kentucky',
       'Mississippi', 'Arkansas', 'Oklahoma', 'Kansas', 'South Dakota',
       'North Dakota', 'Iowa', 'Wisconsin', 'Indiana', 'West Virginia',
       'Maryland', 'Delaware', 'New Jersey', 'Connecticut',
       'Rhode Island', 'Massachusetts', 'Vermont', 'New Hampshire'],
      dtype=object)

In [38]:
df['City'].unique()

array(['New York', 'Houston', 'San Francisco', 'Los Angeles', 'Chicago',
       'Dallas', 'Philadelphia', 'Las Vegas', 'Denver', 'Seattle',
       'Miami', 'Minneapolis', 'Billings', 'Knoxville', 'Omaha',
       'Birmingham', 'Portland', 'Anchorage', 'Honolulu', 'Orlando',
       'Albany', 'Cheyenne', 'Richmond', 'Detroit', 'St. Louis',
       'Salt Lake City', 'New Orleans', 'Boise', 'Phoenix', 'Albuquerque',
       'Atlanta', 'Charleston', 'Charlotte', 'Columbus', 'Louisville',
       'Jackson', 'Little Rock', 'Oklahoma City', 'Wichita',
       'Sioux Falls', 'Fargo', 'Des Moines', 'Milwaukee', 'Indianapolis',
       'Baltimore', 'Wilmington', 'Newark', 'Hartford', 'Providence',
       'Boston', 'Burlington', 'Manchester'], dtype=object)

In [39]:
df['Product'].unique()

array(["Men's Street Footwear", "Men's Athletic Footwear",
       "Women's Street Footwear", "Women's Athletic Footwear",
       "Men's Apparel", "Women's Apparel"], dtype=object)

In [40]:
df['Sales Method'].unique()

array(['In-store', 'Outlet', 'Online'], dtype=object)